ci: add scripting to generate declarative schema changer corpus

fqazi opened the pull request:
These changes do the following:

Informs: #82643

1. Add a test case for validating schema changer corpuses back into master (was accidentally cloberred)
2. Extend logictest to run certain tests in a mixed version test that are interesting from a declarative schema changer viewpoint (we will extend this more)
3. More importantly for the dev-infra team an initial approach to collect declarative schema changer states from logictest during nightly builds. The scripts included here will generate a corpus and upload them to GCS storage for a nightly build.

Release justification: low risk changes to add scripting for nightly artifacts to allow us to do mixed version testing for the declarative schema changer

fqazi committed:
ci: add scripting to generate decl schema changer corpus

Informs: #82643

Previously, we could manually collect data from builds
to do backwards compatibility and mixed version testing
for the declarative schema changer. This was inadequate
because we need continuous testing between branches for
this feature. To address this, this patch will start by
adding infrastructure to run the logictest suite in a
new mode to collect declarative schema changer state
in logictest/logictestccl/ and schema changer tests, and
upload these to GCS.

Release note: None

Release justification:

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86662)
<!-- Reviewable:end -->

fqazi committed:
sql/schemachanger: add test to validate declarative corpus again

Previously, we had a schemachanger test for validating the
corpus file on master that accidentally got clobbered. These
changes add this test back into the corpus subpackage.

Release justification: no real risk adds an new test meant for
nightly CI.
Release note: None

fqazi committed:
sql/logictest: enable mixed version testing for schema changes

Previously, the mixed version tests inside logictests were
used for a single test file. This meant if we used this
to create a corpus of declarative states we wouldn't get
much data. To address this, this patch widens the coverage
to tests that cover schema changes in a targeted manner.

Release justification: low risk test related updates
Release note: None

rickystewart commented on pkg/sql/schemachanger/corpus/BUILD.bazel:
```diff
@@ -18,4 +18,17 @@ go_library(
 
+go_test(
+    name = "corpus_test",
+    srcs = ["corpus_test.go"],
```
I think you should add `tags = ["integration"],` to this `go_test` definition. Otherwise it'll be run as part of the unit test job which I think is not what you want.

rickystewart commented on build/teamcity/cockroach/nightlies/sqllogic_corpus_nightly_impl.sh:
```diff
@@ -0,0 +1,166 @@
+
+# Generate a corpus for all non-mixed version variants
+for config in local multiregion-9node-3region-3azs; do
+$BAZEL_BIN/pkg/cmd/bazci/bazci_/bazci --config=ci \
```
`bazci` is going to get a breaking change to its CLI, so you have to tweak these calls. Please move the `--` from after the target to after `$BAZEL_BIN/pkg/cmd/bazci/bazci_/bazci`, so it'll be like:

```
BAZEL_BIN/pkg/cmd/bazci/bazci_/bazci -- --config=ci
    test //pkg/sql/logictest/tests/$config/... \
    --test_arg=--declarative-corpus=$ARTIFACTS_DIR/corpus \
    --test_env=GO_TEST_WRAP_TESTV=1 \
    --test_env=GO_TEST_WRAP=1 \
    --test_env=GO_TEST_JSON_OUTPUT_FILE=$GO_TEST_JSON_OUTPUT_FILE.$config \
    --test_timeout=7200 \
    || exit_status=$?
```

chengxiong-ruan reviewed (approved):
LGTM on schema stuff.

fqazi commented:
@rickystewart @chengxiong-ruan TFTR!

rickystewart commented:
Oh, please don't check this in until #86959 is merged.

fqazi commented:
Yes, I'll hold off

fqazi commented:
bors r+
